core.js

```diff
@@ -0,0 +1,685 @@
+                                     update.routes,
+                                     update.connections);
+    }
+  }
+  return (clientI >= 0);
+};
+
+/**
+ * Gets an array detailing the clients currently registered with this server.
+ * @returns {Array} An array of client info: 
+ *   [{name:string, description:string, subscribers:Array, 
+ *     publishers:Array, metadata:Object},...]
+ */
+Manager.prototype.getClients = function(){
+  var clients = [];
+  for(var i = 0; i < this.clients.length; i++){
+    var curr = this.clients[i];
+    clients.push(curr.toMap());
+  }
+  return clients;
+};
+
+/**
+ * Gets the index of the first route that matches the provided route.
+ * @param route {Object|string|Route} the route data to match against
+ * @return {number} The index of the first route in the router's
+ *   route list that matches the provided route. 
+ *   Returns -1 if no match is found.
+ */
+Manager.prototype.indexOfRoute = function(route){
+  for(var i = 0; i < this.routes.length; i++){
+    if (this.routes[i].matches(route)){
+      return i;
+    }
+  }
+  return -1;
+};
+
+/**
+ * Goes through all clients and creates any connections that match
+ *   this route.
+ * @param {Route} route The route to create connections with
+ */
+Manager.prototype.connectRoute = function(route){
+  var pubClient,
+      pubClientI;
+  //First we will go through all clients and associate their publishers.
+  // Then we will go through all clients and check their subscribers
+  // against the associated publishers.
+  
+  //for each registered client
+  for(pubClientI = this.clients.length - 1;
+      pubClientI >= 0;
+      pubClientI--){
+    pubClient = this.clients[pubClientI];
+    //connect this client's publishers to all 
+    // appropriate registered subscribers
+    this.connectRoutePubClient(route, pubClient);
+  }
+};
+
+/**
+ * Adds a route to the server
+ * @param route {Route} The route object to add. 
+ * @returns {boolean} true iff the provided route is added. 
+ *   The route will not be added if it matches a route already registered.
+ * @throws Error if argument is not a Route instance
+ */
+Manager.prototype.addRoute = function(route){
+  if (!(route instanceof Route)){
+    throw new TypeError('argument must be a Route instance');
+  }
+  var routeUnique = (this.indexOfRoute(route) === -1);
+  if (routeUnique){
+    this.connectRoute(route);
+    this.routes.push(route);
+
+    //get all the connections this route is associated with
+    var connections = route.getConnections();
+    //filter out any connections that were 
+    //previously-defined by other routes
+    connections = connections.filter(
+      function(element/*, index, array*/){
+        return (element.routes.length === 1 &&
+                element.routes[0] === route.uuid);
+      });
+
+    //update the admins
+    for(var adminI = this.admins.length - 1; adminI >= 0; adminI--){
+      this.admins[adminI].sendAdd(undefined,
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Unexpected use of undefined.](https://www.codacy.com/app/quinkennedy/spacebrew/file/2279533613/issues/source?bid=3184377&fileBranchId=3187330#l505)